README.md
Vianetz Attach Terms And Conditions Extension for Magento
=========================================================

Facts
-----
- extension key: Vianetz_AttachTermsAndConditions

Description
-----------
This module for the Magento 1 online shop software offers the possibility to attach PDF documents
(e.g. Terms And Conditions) to add to Magento order confirmation emails.

Requirements
------------
- PHP >= 5.6.0
- [Vianetz_PdfAttachments](https://github.com/vianetz/magento1-pdfattachments)

Compatibility
-------------
- Magento >= 1.7

Usage
-----
This extension loops through all agreements that have been created in Magento Admin Sales > Terms and Conditions and checks
if a PDF file exists in the _media/_ folder of the Magento directory with exactly the name of the agreement.
In that case this file is attached to the sales order email.

For example if the agreement in the backend has the name "AGB" then the extension looks for the file _media/AGB.pdf_
and if this file exists it is attached to the sales order transactional email.

Installation Instructions
-------------------------
For installation notes please see also [our FAQ](https://www.vianetz.com/en/faq/how-to-install-the-magento-extension.html).

1. Do a backup of your Magento installation for safety reasons.
2. Disable Magento compilation feature (if activated): System > Tools > Compiler
3. Unzip the setup package and copy the contents of the src/ folder into the Magento root folder. (The folder structure
   is the same as in your Magento installation. No files will be overwritten.)
   Please assure that the files are uploaded with the same file user permissions as the Magento installation!
4. Clear the Magento cache (and related caches like APC if available)
5. Logout from the admin panel and then login again
6. Enable the Magento compilation feature (if it was activated before): System > Tools > Compiler

As an alternative you can install the module via modman.
Please find more information about that installation method at [https://github.com/colinmollenhour/modman](https://github.com/colinmollenhour/modman)
(Thanks @colinmollenhour)

We also offer paid installation services. If you are interested please contact me.

Uninstallation
--------------
1. Remove the folder `app/code/community/Vianetz/AttachTermsAndConditions`
2. Remove the file `app/etc/modules/Vianetz_AttachTermsAndConditions.xml`

Frequently Asked Questions
--------------------------
Please find the [Frequently Asked Questions on our website](https://www.vianetz.com/en/faq).

Support
-------
If you have any issues or suggestions with this extension, please do not hesitate to 
[contact me](https://www.vianetz.com/en/contacts).

Developer
---------
Christoph Massmann
[https://www.vianetz.com](https://www.vianetz.com)
[@vianetz](https://twitter.com/vianetz)

Licence
-------
[GNU GENERAL PUBLIC LICENSE](http://www.gnu.org/licenses/gpl-3.0.txt)

Copyright
---------
(c) since 2006 vianetz

This Magento Extension uses Semantic Versioning - please find more information at http://semver.org.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [2.2.1] - 2021-02-17
### Fixed
- Fixed issue with PHP 8

## [2.2.0] - 2020-04-16
### Added
- Added `composer.json`
### Removed
- Removed dependency to `Vianetz_Core` extension

## [2.1.0] - 2018-12-09
### Added
- Added configuration flag to specify whether to filter for active agreements or not

## [2.0.4] - 2018-09-17
### Added
- Release module under GPL


app/code/local/Vianetz/AttachTermsAndConditions/Model/Order.php
<?php
/**
 * AttachTermsAndConditions Invoice Class
 *
 * @category Vianetz
 * @package AttachTermsAndConditions
 * @author Christoph Massmann <C.Massmann@vianetz.com>
 */
class Vianetz_AttachTermsAndConditions_Model_Order extends Mage_Sales_Model_Order
{
    /**
     * Sending email with order data
     *
     * @return Mage_Sales_Model_Order
     */
    public function sendNewOrderEmail()
    {
        if (!Mage::helper('sales')->canSendNewOrderEmail($this->getStore()->getId())) {
            return $this;
        }

        $translate = Mage::getSingleton('core/translate');
        /* @var $translate Mage_Core_Model_Translate */
        $translate->setTranslateInline(false);

        $paymentBlock = Mage::helper('payment')->getInfoBlock($this->getPayment())
            ->setIsSecureMode(true);

        $paymentBlock->getMethod()->setStore($this->getStore()->getId());

        $mailTemplate = Mage::getModel('core/email_template');
        /* @var $mailTemplate Mage_Core_Model_Email_Template */
        $copyTo = $this->_getEmails(self::XML_PATH_EMAIL_COPY_TO);
        $copyMethod = Mage::getStoreConfig(self::XML_PATH_EMAIL_COPY_METHOD, $this->getStoreId());
        if ($copyTo && $copyMethod == 'bcc') {
            foreach ($copyTo as $email) {
                $mailTemplate->addBcc($email);
            }
        }

        if ($this->getCustomerIsGuest()) {
            $template = Mage::getStoreConfig(self::XML_PATH_EMAIL_GUEST_TEMPLATE, $this->getStoreId());
            $customerName = $this->getBillingAddress()->getName();
        } else {
            $template = Mage::getStoreConfig(self::XML_PATH_EMAIL_TEMPLATE, $this->getStoreId());
            $customerName = $this->getCustomerName();
        }

        $sendTo = array(
            array(
                'email' => $this->getCustomerEmail(),
                'name'  => $customerName
            )
        );
        if ($copyTo && $copyMethod == 'copy') {
            foreach ($copyTo as $email) {
                $sendTo[] = array(
                    'email' => $email,
                    'name'  => null
                );
            }
        }

        if (Mage::getStoreConfigFlag('checkout/options/enable_agreements')) {
            $agreements = Mage::getModel('checkout/agreement')->getCollection()
                ->addStoreFilter(Mage::app()->getStore()->getId())
                ->addFieldToFilter('is_active', 1);

            foreach ( $agreements as $agreement ) {
                $baseDir = Mage::getBaseDir('media');
                $file = $baseDir . '/' . $agreement->getName() . '.pdf';
                if ( file_exists($file)) {
                    $mailTemplate->addAttachment(file_get_contents($file), Mage::helper('sales')->__($agreement->getName()).'.pdf');
                }
            }
        }

        foreach ($sendTo as $recipient) {
            $mailTemplate->setDesignConfig(array('area'=>'frontend', 'store'=>$this->getStoreId()))
                ->sendTransactional(
                $template,
                Mage::getStoreConfig(self::XML_PATH_EMAIL_IDENTITY, $this->getStoreId()),
                $recipient['email'],
                $recipient['name'],
                array(
                     'order'         => $this,
                     'billing'       => $this->getBillingAddress(),
                     'payment_html'  => $paymentBlock->toHtml(),
                )
            );
        }

        $this->setEmailSent(true);
        $this->_getResource()->saveAttribute($this, 'email_sent');

        $translate->setTranslateInline(true);

        return $this;
    }
}


src/app/code/community/Vianetz/AttachTermsAndConditions/Helper/Data.php
<?php
/**
 * AttachTermsAndConditions Helper Class
 *
 * @section LICENSE
 * This file is created by vianetz <info@vianetz.com>.
 * The code is distributed under the GPL license.
 *
 * If you did not receive a copy of the license and are unable to
 * obtain it through the world-wide-web, please send an email
 * to license@vianetz.com so we can send you a copy immediately.
 *
 * @category    Vianetz
 * @package     Vianetz\AttachTermsAndConditions
 * @author      Christoph Massmann, <C.Massmann@vianetz.com>
 * @link        https://www.vianetz.com
 * @copyright   Copyright (c) since 2006 vianetz - Dipl.-Ing. C. Massmann (https://www.vianetz.com)
 * @license     http://www.gnu.org/licenses/gpl-3.0.txt GNU GENERAL PUBLIC LICENSE
 */
class Vianetz_AttachTermsAndConditions_Helper_Data extends Mage_Core_Helper_Abstract
{
    /**
     * Log message to file if enabled in system configuration.
     *
     * @param string $message
     * @param int $type
     *
     * @return Vianetz_AttachTermsAndConditions_Helper_Data
     */
    public function log($message, $type = LOG_DEBUG)
    {
        $extensionNamespace = 'Vianetz_AttachTermsAndConditions';

        $extensionVersion = Mage::getConfig()->getModuleConfig($extensionNamespace)->version;
        $message = $extensionNamespace . ' v' . $extensionVersion . ': ' . $message;
        $logFilename = $extensionNamespace . '.log';

        Mage::log($message, $type, $logFilename, true);

        return $this;
    }
}

src/app/code/community/Vianetz/AttachTermsAndConditions/Model/Observer.php
<?php
/**
 * AttachTermsAndConditions Email Observer
 *
 * @section LICENSE
 * This file is created by vianetz <info@vianetz.com>.
 * The code is distributed under the GPL license.
 *
 * If you did not receive a copy of the license and are unable to
 * obtain it through the world-wide-web, please send an email
 * to license@vianetz.com so we can send you a copy immediately.
 *
 * @category    Vianetz
 * @package     Vianetz\AttachTermsAndConditions
 * @author      Christoph Massmann, <C.Massmann@vianetz.com>
 * @link        https://www.vianetz.com
 * @copyright   Copyright (c) since 2006 vianetz - Dipl.-Ing. C. Massmann (https://www.vianetz.com)
 * @license     http://www.gnu.org/licenses/gpl-3.0.txt GNU GENERAL PUBLIC LICENSE
 */
class Vianetz_AttachTermsAndConditions_Model_Observer
{
    /**
     * Add pdf documents as email attachments.
     *
     * Event: vianetz_pdfattachments_email_template_init
     *
     * @param Varien_Event_Observer $observer
     *
     * @return Vianetz_AttachTermsAndConditions_Model_Observer
     */
    public function addPdfAttachment(Varien_Event_Observer $observer)
    {
        try {
            /** @var Mage_Core_Model_Email_Template_Mailer $mailer */
            $mailer = $observer->getEvent()->getMailer();

            $orderEmailTemplateId = Mage::getStoreConfig(Mage_Sales_Model_Order::XML_PATH_EMAIL_TEMPLATE, $mailer->getStoreId());
            $orderEmailGuestTemplateId = Mage::getStoreConfig(Mage_Sales_Model_Order::XML_PATH_EMAIL_GUEST_TEMPLATE, $mailer->getStoreId());

            if ($mailer->getTemplateId() != $orderEmailTemplateId && $mailer->getTemplateId() != $orderEmailGuestTemplateId) {
                return $this;
            }

            /** @var Mage_Core_Model_Email_Template $emailTemplate */
            $emailTemplate = $observer->getEvent()->getEmailTemplate();

            foreach ($this->getAgreements() as $agreement) {
                $file = Mage::getBaseDir('media') . DS . $agreement->getName() . '.pdf';
                $this->getHelper()->log('Searching for attachment file ' . $file);

                if (file_exists($file) === true) {
                    $orderId = ($this->getOrder() !== null) ? $this->getOrder()->getIncrementId() : '';
                    $this->getHelper()->log('Attaching ' . $file . ' to order email for order #' . $orderId . '.');

                    $filename = Mage::helper('sales')->__($agreement->getName()) . '.pdf';
                    Mage::helper('pdfattachments')->addAttachmentToEmail($emailTemplate, file_get_contents($file), $filename);
                }
            }
        } catch (Exception $exception) {
            $this->getHelper()->log('Error while attaching pdf document: ' . $exception->getMessage() . ' ' . $exception->getTraceAsString());
        }

        return $this;
    }

    /**
     * @return int
     * @throws \Mage_Core_Model_Store_Exception
     */
    private function getStoreId()
    {
        $storeId = (int)Mage::app()->getStore()->getId();

        if ($storeId === Mage_Core_Model_App::ADMIN_STORE_ID && $this->getOrder() !== null) {
            $storeId = $this->getOrder()->getStoreId();
        }

        return $storeId;
    }

    /**
     * @return Mage_Sales_Model_Order|null
     */
    private function getOrder()
    {
        return Mage::registry('current_order');
    }

    /**
     * @return \Vianetz_AttachTermsAndConditions_Helper_Data
     */
    private function getHelper()
    {
        return Mage::helper('attachtermsandconditions');
    }

    /**
     * @return \Mage_Checkout_Model_Resource_Agreement_Collection
     * @throws \Mage_Core_Model_Store_Exception
     */
    private function getAgreements()
    {
        $this->getHelper()->log('Looking for all agreements in store ' . $this->getStoreId());
        $agreements = Mage::getModel('checkout/agreement')->getCollection()
            ->addStoreFilter($this->getStoreId());

        $isFilterActiveAgreements = Mage::getStoreConfigFlag('vianetz_attachtermsandconditions/general/filter_active_agreements', $this->getStoreId());
        if ($isFilterActiveAgreements) {
            $agreements->addFieldToFilter('is_active', 1);
        }

        if (count($agreements) === 0) {
            $this->getHelper()->log('No active agreements found for store ' . $this->getStoreId() . '.');
        }

        return $agreements;
    }
}

app/code/local/Vianetz/AttachTermsAndConditions/Model/Order/Invoice.php
<?php
/**
 * AttachTermsAndConditions Invoice Class
 * 
 * @category Vianetz
 * @package AttachTermsAndConditions
 * @author Christoph Massmann <C.Massmann@vianetz.com>
 */
class Vianetz_AttachTermsAndConditions_Model_Order_Invoice extends
Mage_Sales_Model_Order_Invoice
{  
    /**
     * Sending email with Invoice data
     *
     * @param bool $notifyCustomer
     * @param string $comment
     * @return Mage_Sales_Model_Order_Invoice
     */
    public function sendEmail($notifyCustomer=true, $comment='')
    {
        if (!Mage::helper('sales')->canSendNewInvoiceEmail($this->getOrder()->getStore()->getId())) {
            return $this;
        }

        $currentDesign = Mage::getDesign()->setAllGetOld(array(
            'package' => Mage::getStoreConfig('design/package/name', $this->getStoreId()),
            'store'   => $this->getStoreId()
        ));

        $translate = Mage::getSingleton('core/translate');
        /* @var $translate Mage_Core_Model_Translate */
        $translate->setTranslateInline(false);

        $order  = $this->getOrder();
        $copyTo = $this->_getEmails(self::XML_PATH_EMAIL_COPY_TO);
        $copyMethod = Mage::getStoreConfig(self::XML_PATH_EMAIL_COPY_METHOD, $this->getStoreId());

        if (!$notifyCustomer && !$copyTo) {
            return $this;
        }
        $paymentBlock   = Mage::helper('payment')->getInfoBlock($order->getPayment())
            ->setIsSecureMode(true);

        $mailTemplate = Mage::getModel('core/email_template');

        if ($order->getCustomerIsGuest()) {
            $template = Mage::getStoreConfig(self::XML_PATH_EMAIL_GUEST_TEMPLATE, $order->getStoreId());
            $customerName = $order->getBillingAddress()->getName();
        } else {
            $template = Mage::getStoreConfig(self::XML_PATH_EMAIL_TEMPLATE, $order->getStoreId());
            $customerName = $order->getCustomerName();
        }

        if ($notifyCustomer) {
            $sendTo[] = array(
                'name'  => $customerName,
                'email' => $order->getCustomerEmail()
            );
            if ($copyTo && $copyMethod == 'bcc') {
                foreach ($copyTo as $email) {
                    $mailTemplate->addBcc($email);
                }
            }

        }

        if ($copyTo && ($copyMethod == 'copy' || !$notifyCustomer)) {
            foreach ($copyTo as $email) {
                $sendTo[] = array(
                    'name'  => null,
                    'email' => $email
                );
            }
        }

        // BEGIN Customization by vianetz
        // 2009-04-17
            if (Mage::getStoreConfigFlag('checkout/options/enable_agreements')) {
                $agreements = Mage::getModel('checkout/agreement')->getCollection()
                    ->addStoreFilter(Mage::app()->getStore()->getId())
                    ->addFieldToFilter('is_active', 1);
#Mage::log($agreements);

        foreach ( $agreements as $agreement ) {
Mage::log($agreement);

$pdf = new Zend_Pdf();
         $style = new Zend_Pdf_Style();

      $font = Zend_Pdf_Font::fontWithName(Zend_Pdf_Font::FONT_HELVETICA);
       $style->setFont($font, 10);


Mage::log("TEST..");
Mage::log($agreement->getName());

$page = $pdf->newPage(Zend_Pdf_Page::SIZE_A4);
$pdf->pages[] = $page;
$page->setFillColor(new Zend_Pdf_Color_GrayScale(0));
$page->setFont($font, 10);

Mage::log("TEST...");
$page->drawText("Allgemeine Geschaeftsbedingungen", 89, 700, 'UTF-8');
$text = wordwrap($agreement->getContent(), 70, "\n", false);
    $text = explode("\n", $text);
$y = 720;
    foreach ($text as $row) {
        if ($y < 60) {
            $page = $pdf->newPage(Zend_Pdf_Page::SIZE_A4);
            $pdf->pages[] = $page;
            $page->setFont(Zend_Pdf_Font::fontWithName(Zend_Pdf_Font::FONT_HELVETICA),
12);
            $y = 760;
        } 
        $page->drawText($row, 89, $y);
        $y-=15;
    }

Mage::log("TEST....");
            $mailTemplate->addAttachment($pdf,
Mage::helper('sales')->__($agreement->getName().'.pdf'));
                break;
        }


        }

        // END Customization by vianetz

        foreach ($sendTo as $recipient) {
            $mailTemplate->setDesignConfig(array('area'=>'frontend', 'store'=>$order->getStoreId()))
                ->sendTransactional(
                    $template,
                    Mage::getStoreConfig(self::XML_PATH_EMAIL_IDENTITY, $order->getStoreId()),
                    $recipient['email'],
                    $recipient['name'],
                    array(
                        'order'       => $order,
                        'invoice'     => $this,
                        'comment'     => $comment,
                        'billing'     => $order->getBillingAddress(),
                        'payment_html'=> $paymentBlock->toHtml(),
                    )
                );
        }

        $translate->setTranslateInline(true);

        Mage::getDesign()->setAllGetOld($currentDesign);

        return $this;
    }
}

/* vim: set ts=4 sw=4 expandtab nu tw=90: */
